File: apps/server-admin/src/lib/system-notifications/providers/magicpush.ts
import type {
  NotificationMessage,
  NotificationProvider,
  NotificationProviderDefinition,
  NotificationSchemaField,
  NotificationSendResult,
} from "../types";
import { toPlainRecord, toTrimmedString, truncateText } from "./shared";
import { tDefault } from "../../i18n";

const magicpushT = (
  key: string,
  params?: Record<string, string | number | boolean | null | undefined>,
) =>
  tDefault(`server.notifications.providers.catalog.magicpush.${key}`, params);

const MAGICPUSH_CONNECTION_SCHEMA: NotificationSchemaField[] = [
  {
    key: "server_url",
    label: magicpushT("fields.server_url.label"),
    description: magicpushT("fields.server_url.description"),
    placeholder: "http://192.168.31.98:3000",
    type: "string",
    required: true,
  },
  {
    key: "delivery_mode",
    label: magicpushT("fields.delivery_mode.label"),
    description: magicpushT("fields.delivery_mode.description"),
    type: "select",
    required: true,
    default_value: "push",
    options: [
      { label: magicpushT("fields.delivery_mode.options.push"), value: "push" },
      {
        label: magicpushT("fields.delivery_mode.options.inbound"),
        value: "inbound",
      },
    ],
  },
  {
    key: "token",
    label: "Token",
    description: magicpushT("fields.token.description"),
    placeholder: "your_token",
    type: "string",
    required: true,
  },
  {
    key: "timeout_seconds",
    label: magicpushT("fields.timeout_seconds.label"),
    type: "number",
    required: true,
    default_value: 5,
    min: 1,
    max: 30,
  },
];

const MAGICPUSH_TARGET_SCHEMA: NotificationSchemaField[] = [];

export const magicpushProviderDefinition: NotificationProviderDefinition = {
  type: "magicpush",
  label: magicpushT("label"),
  description: magicpushT("description"),
  connection_schema: MAGICPUSH_CONNECTION_SCHEMA,
  target_schema: MAGICPUSH_TARGET_SCHEMA,
  sensitive_fields: [],
  capabilities: {
    supports_text: true,
    supports_markdown: false,
    supports_rich_blocks: false,
    supports_actions: false,
    supports_mentions: false,
    supports_attachments: false,
    supports_provider_dedupe_key: false,
    max_body_length: null,
  },
};

const resolveMagicPushUrl = (
  provider: NotificationProvider,
  token: string,
  deliveryMode: "push" | "inbound",
) => {
  const baseUrl = toTrimmedString(provider.connection_config.server_url);
  if (!baseUrl) {
    throw new Error(magicpushT("errors.missingBaseUrl"));
  }

  const normalizedBaseUrl = baseUrl.replace(/\/+$/, "");
  if (deliveryMode === "inbound") {
    if (/\/api\/inbound\/[^/]+$/i.test(normalizedBaseUrl)) {
      return normalizedBaseUrl;
    }
    if (/\/api\/inbound$/i.test(normalizedBaseUrl)) {
      return `${normalizedBaseUrl}/${encodeURIComponent(token)}`;
    }
    return `${normalizedBaseUrl}/api/inbound/${encodeURIComponent(token)}`;
  }

  if (/\/api\/push(?:\/[^/]+)?$/i.test(normalizedBaseUrl)) {
    return normalizedBaseUrl;
  }

  return `${normalizedBaseUrl}/api/push`;
};

const buildMagicPushContent = (message: NotificationMessage) => {
  const sections: string[] = [];

  if (message.summary?.trim()) {
    sections.push(message.summary.trim());
  }

  if (message.body_text?.trim()) {
    sections.push(message.body_text.trim());
  }

  if (message.facts.length > 0) {
    sections.push(
      message.facts.map((fact) => `${fact.label}: ${fact.value}`).join("\n"),
    );
  }

  if (message.actions.length > 0) {
    sections.push(
      message.actions
        .filter(
          (action) =>
            toTrimmedString(action.label) && toTrimmedString(action.url),
        )
        .map((action) => `${action.label}: ${action.url}`)
        .join("\n"),
    );
  }

  return sections.filter(Boolean).join("\n\n");
};

const parseMagicPushApiCode = (response: Record<string, unknown> | null) => {
  const rawCode = response?.code;
  if (typeof rawCode === "number") {
    return rawCode;
  }

  const parsed = Number.parseInt(String(rawCode ?? ""), 10);
  return Number.isFinite(parsed) ? parsed : undefined;
};

const buildMagicPushInboundPayload = (message: NotificationMessage) => {
  const content = buildMagicPushContent(message) || message.title;
  const facts = message.facts.reduce<Record<string, string>>((acc, fact) => {
    const label = toTrimmedString(fact.label);
    if (!label) return acc;
    acc[label] = fact.value;
    return acc;
  }, {});

  return {
    source: "fn-knock",
    title: toTrimmedString(
      message.title || message.summary || magicpushT("message.fallbackTitle"),
    ),
    summary: message.summary,
    content,
    body: content,
    body_text: message.body_text,
    body_markdown: message.body_markdown,
    type: message.body_markdown ? "markdown" : "text",
    severity: message.severity,
    facts,
    facts_list: message.facts,
    actions: message.actions,
    mentions: message.mentions,
    dedupe_key: message.dedupe_key,
    occurred_at: message.occurred_at,
    event_id: message.event_id,
    metadata: message.metadata,
  };
};

export const sendMagicPushMessage = async (args: {
  provider: NotificationProvider;
  message: NotificationMessage;
  timeoutSeconds: number;
}): Promise<NotificationSendResult> => {
  const token = toTrimmedString(args.provider.connection_config.token);
  if (!token) {
    return {
      success: false,
      retryable: false,
      reason: magicpushT("errors.missingToken"),
    };
  }

  const deliveryMode =
    toTrimmedString(args.provider.connection_config.delivery_mode) === "inbound"
      ? "inbound"
      : "push";
  let url: string;
  try {
    url = resolveMagicPushUrl(args.provider, token, deliveryMode);
  } catch (error) {
    return {
      success: false,
      retryable: false,
      reason:
        error instanceof Error ? error.message : magicpushT("errors.invalidBaseUrl"),
    };
  }

  const title = toTrimmedString(
    args.message.title ||
      args.message.summary ||
      magicpushT("message.fallbackTitle"),
  );
  const content = buildMagicPushContent(args.message) || title;
  const payload =
    deliveryMode === "inbound"
      ? buildMagicPushInboundPayload(args.message)
      : {
          title,
          content,
          type: "text",
        };
  const headers: Record<string, string> =
    deliveryMode === "inbound"
      ? {
          "content-type": "application/json; charset=utf-8",
          "x-fn-knock-provider": "magicpush",
        }
      : {
          authorization: `Bearer ${token}`,
          "content-type": "application/json; charset=utf-8",
        };

  const controller = new AbortController();
  const timeout = setTimeout(
    () => controller.abort(),
    Math.max(1, args.timeoutSeconds) * 1000,
  );

  try {
    const response = await fetch(url, {
      method: "POST",
      headers,
      body: JSON.stringify(payload),
      signal: controller.signal,
    });
    const responseText = await response.text().catch(() => "");
    let parsedResponse: Record<string, unknown> | null = null;
    try {
      parsedResponse = responseText
        ? (JSON.parse(responseText) as Record<string, unknown>)
        : null;
    } catch {
      parsedResponse = null;
    }

    const apiCode = parseMagicPushApiCode(parsedResponse);
    const apiMessage = toTrimmedString(
      parsedResponse?.message ?? parsedResponse?.msg ?? parsedResponse?.error,
    );
    const data = toPlainRecord(parsedResponse?.data);
    const apiSuccess = parsedResponse?.success;
    const succeeded =
      response.ok &&
      apiSuccess !== false &&
      (apiCode === undefined || apiCode === 200);

    return {
      success: succeeded,
      retryable:
        !succeeded && (response.status >= 500 || response.status === 429),
      reason: succeeded
        ? undefined
        : apiMessage ||
          magicpushT("errors.requestReturned", { status: response.status }),
      request_summary: {
        method: "POST",
        url,
        delivery_mode: deliveryMode,
        type: payload.type,
        title_preview: payload.title,
        content_preview: truncateText(payload.content),
      },
      response_summary: {
        status: response.status,
        ok: response.ok,
        success: apiSuccess,
        code: apiCode,
        message: apiMessage || undefined,
        total: data.total,
        success_count: data.successCount,
        failed_count: data.failedCount,
        body_preview: truncateText(responseText),
      },
    };
  } catch (error) {
    return {
      success: false,
      retryable: true,
      reason:
        error instanceof Error
          ? error.message
          : magicpushT("errors.requestFailed"),
      request_summary: {
        method: "POST",
        url,
        delivery_mode: deliveryMode,
        type: payload.type,
        title_preview: payload.title,
      },
    };
  } finally {
    clearTimeout(timeout);
  }
};

File: apps/server-admin-rs/src/notifications/routes/providers/magicpush.rs
use super::*;

pub(in crate::notifications::routes) fn magicpush_definition() -> ProviderDefinition {
    ProviderDefinition {
        provider_type: "magicpush",
        label: "MagicPush",
        description: "Send notifications through MagicPush.",
        connection_schema: vec![
            string_schema("server_url", "Server URL", true, false, None)
                .placeholder("http://192.168.31.98:3000"),
            select_schema(
                "delivery_mode",
                "Delivery mode",
                true,
                Some("push"),
                &["push", "inbound"],
            ),
            string_schema("token", "Token", true, false, None).placeholder("your_token"),
            number_schema("timeout_seconds", "Timeout seconds", true, Some(5)).bounds(1, 30),
        ],
        target_schema: Vec::new(),
        sensitive_fields: Vec::new(),
        supports_markdown: false,
        supports_actions: false,
        supports_mentions: false,
        supports_provider_dedupe_key: false,
    }
}

pub(in crate::notifications::routes) async fn send_magicpush(
    state: &AppState,
    provider: &Value,
    message: &Value,
    timeout_seconds: i64,
) -> ProviderTestResult {
    let config = provider_config(provider);
    let token = config_text(&config, "token");
    if token.is_empty() {
        return missing_config_result(&notification_provider_error_default(
            "magicpush",
            "missingToken",
            &[],
        ));
    }
    let base_url = config_text(&config, "server_url");
    if base_url.is_empty() {
        return missing_config_result(&notification_provider_error_default(
            "magicpush",
            "missingBaseUrl",
            &[],
        ));
    }
    let delivery_mode = if config_text(&config, "delivery_mode") == "inbound" {
        "inbound"
    } else {
        "push"
    };
    let url = resolve_magicpush_url(&base_url, &token, delivery_mode);
    let title = message_title(message);
    let content = default_string(build_magicpush_content(message), &title);
    let magicpush_facts = magicpush_facts_object(message);
    let payload = if delivery_mode == "inbound" {
        json!({
            "source": "fn-knock",
            "title": title,
            "summary": message.get("summary").cloned().unwrap_or(Value::Null),
            "content": content,
            "body": content,
            "body_text": message.get("body_text").cloned().unwrap_or(Value::Null),
            "body_markdown": message.get("body_markdown").cloned().unwrap_or(Value::Null),
            "type": if message_text(message, "body_markdown").is_empty() { "text" } else { "markdown" },
            "severity": message.get("severity").cloned().unwrap_or(Value::Null),
            "facts": magicpush_facts,
            "facts_list": message.get("facts").cloned().unwrap_or_else(|| json!([])),
            "actions": message.get("actions").cloned().unwrap_or_else(|| json!([])),
            "mentions": message.get("mentions").cloned().unwrap_or_else(|| json!([])),
            "dedupe_key": message.get("dedupe_key").cloned().unwrap_or(Value::Null),
            "occurred_at": message.get("occurred_at").cloned().unwrap_or(Value::Null),
            "event_id": message.get("event_id").cloned().unwrap_or(Value::Null),
            "metadata": message.get("metadata").cloned().unwrap_or_else(|| json!({}))
        })
    } else {
        json!({ "title": title, "content": content, "type": "text" })
    };
    let request_summary = json!({
        "method": "POST",
        "url": url,
        "delivery_mode": delivery_mode,
        "type": payload.get("type").cloned().unwrap_or(Value::Null),
        "title_preview": payload.get("title").cloned().unwrap_or(Value::Null),
        "content_preview": truncate_text(payload.get("content").and_then(Value::as_str).unwrap_or(""), 500)
    });
    let mut request = state
        .fallback_client
        .post(&url)
        .header("content-type", "application/json; charset=utf-8");
    if delivery_mode == "inbound" {
        request = request.header("x-fn-knock-provider", "magicpush");
    } else {
        request = request.header("authorization", format!("Bearer {token}"));
    }
    let (status, ok, text, parsed) = send_prepared_json(request, &payload, timeout_seconds).await;
    provider_result_from_api(
        "MagicPush",
        request_summary,
        status,
        ok,
        text,
        parsed,
        |value| {
            value.get("success").and_then(Value::as_bool) != Some(false)
                && json_i64(value, "code").is_none_or(|code| code == 200)
        },
        |value| json_text_any(value, &["message", "msg", "error"]),
    )
}
